server/src/main/java/org/opensearch/action/search/SearchRequestListenerManager.java

```diff
@@ -0,0 +1,125 @@
+    ) {
+        this.clusterService = clusterService;
+        searchRequestListenersList = new ArrayList<>();
+    }
+
+    /**
+     * Add multiple {@link SearchRequestOperationsListener} to the searchRequestListenersList.
+     * Those enabled listeners will be executed during each search request.
+     *
+     * @param listeners Multiple SearchRequestOperationsListener object to add.
+     * @throws IllegalArgumentException if any input listener is null or already exists in the list.
+     */
+    public void addListeners(SearchRequestOperationsListener... listeners) {
+        for (SearchRequestOperationsListener listener : listeners) {
+            if (listener == null) {
+                throw new IllegalArgumentException("listener must not be null");
+            }
+            if (searchRequestListenersList.contains(listener)) {
+                throw new IllegalArgumentException("listener already added");
+            }
+            searchRequestListenersList.add(listener);
+        }
+    }
+
+    /**
+     * Remove a {@link SearchRequestOperationsListener} from the searchRequestListenersList,
+     *
+     * @param listener A SearchRequestOperationsListener object to remove.
+     * @throws IllegalArgumentException if the input listener is null or already exists in the list.
+     */
+    public void removeListener(SearchRequestOperationsListener listener) {
+        if (listener == null) {
+            throw new IllegalArgumentException("listener must not be null");
+        }
+        if (!searchRequestListenersList.contains(listener)) {
+            throw new IllegalArgumentException("listener does not exist in the listeners list");
+        }
+        searchRequestListenersList.remove(listener);
+    }
+
+    /**
+     * Get searchRequestListenersList,
+     *
+     * @return List of SearchRequestOperationsListener
+     * @throws IllegalArgumentException if the input listener is null or already exists in the list.
+     */
+    public List<SearchRequestOperationsListener> getListeners() {
+        return searchRequestListenersList;
+    }
+
+
+    /**
+     * Create the {@link SearchRequestOperationsListener.CompositeListener}
+     * with the all listeners enabled at cluster-level and request-level.
+     *
+     * @param searchRequest The SearchRequest object. SearchRequestListenerManager will decide which request-level listeners to add based on states/flags of the request
+     * @param logger Logger to be attached to the {@link SearchRequestOperationsListener.CompositeListener}
+     * @param perRequestListeners the per-request listeners that can be optionally added to the returned CompositeListener list.
+     * @return SearchRequestOperationsListener.CompositeListener
+     */
+    public SearchRequestOperationsListener.CompositeListener buildCompositeListener(
+        SearchRequest searchRequest,
+        Logger logger,
+        SearchRequestOperationsListener... perRequestListeners
+    ) {
```

Comment:
Can you change the `isEnabled` method on `SearchRequestOperationsListener` to take the `SearchRequest`? 

Then you can override the `isEnabled` method in `TimeProvider` to return `true` if the cluster setting is true or if `searchRequest.isPhaseTook()` is true. Then you can apply consistent logic where you return a list of all enabled listeners.

Reply:
This is actually a very good point to have the logic to decide if a listener should be enabled to be self-contained in each listener!